### ISSUE TITLE
Issue with import of node.js based wikis with plugins loaded via tiddlywiki.info

### ISSUE THREAD

**Issue by mauloop** [OokTech/TW5-Bob#81]
My environment is: Win10 + Chrome 72 + BobExe 1.2.1.

I found a way to avoid the behaviour I am going to describe, therefore I'm ok, this is not blocking for my work. Anyway I'd like to report it, since you may have a better solution to suggest. If not, these notes may help to fix the issue and/or update the docs.

What did I do? 

I had a wiki running under node.js, not Bob, (let's call it "_mywiki_"), which made use of a couple of plugins loaded via `tiddlywiki.info`.

1. I copied my plugins into `Bob/Plugins` folder, ... 
2. verified that plugins are correctly listed in "_Control Panel/Bob Settings/Manage Plugins_" tab, ...
3. stopped old node.js instance, ...
4. updated Bob settings with the path of my existing wiki, ...
5. restarted Bob ...
6. and finally opened _mywiki_ with the url provided by Bob (`127.0.0.1:8080/mywiki`)

While opening the wiki lot of alerts were displayed in quick succession. The following is the content of the alerts (all of them were titled "_syncer-browser-tiddlyweb_"):

```
XMLHttpRequest error code: 404
Sync error while processing '$:/StoryList': XMLHttpRequest error code: 404
Sync error while processing '$:/plugins/OokTech/Bob/Unsent': XMLHttpRequest error code: 404
Sync error while processing '$:/Bob/VisibleConfigurationTabs': XMLHttpRequest error code: 404
Sync error while processing '$:/WikiSettings/split': XMLHttpRequest error code: 404
Sync error while processing '$:/WikiSettings/split/excludePluginList': XMLHttpRequest error code: 404
Sync error while processing '$:/WikiSettings/split/includePluginList': XMLHttpRequest error code: 404
Sync error while processing '$:/WikiSettings/split/serverInfo': XMLHttpRequest error code: 404
Sync error while processing '$:/WikiSettings/split/API': XMLHttpRequest error code: 404
Sync error while processing '$:/WikiSettings/split/ws-server': XMLHttpRequest error code: 404
Sync error while processing '$:/WikiSettings/split/wikis': XMLHttpRequest error code: 404
Sync error while processing '$:/WikiSettings': XMLHttpRequest error code: 404
Sync error while processing '$:/ServerIP': XMLHttpRequest error code: 404
Sync error while processing '$:/EditionsList': XMLHttpRequest error code: 404
Sync error while processing '$:/Bob/ActiveLanguagesList': XMLHttpRequest error code: 404
Sync error while processing '$:/Bob/ActiveThemesList': XMLHttpRequest error code: 404
Sync error while processing '$:/Bob/ActivePluginList': XMLHttpRequest error code: 404
```
In the meantime these error messages were logged to the browser console:

```
status:1 Failed to load resource: the server responded with a status of 404 (Not Found)
EditionM:1 Failed to load resource: the server responded with a status of 404 (Not Found)
EditionM:1 Failed to load resource: the server responded with a status of 404 (Not Found)
recipes/undefined/tiddlers/%24%3A%2FStoryList:1 Failed to load resource: the server responded with a status of 404 (Not Found)
recipes/undefined/tiddlers/%24%3A%2Fplugins%2FOokTech%2FBob%2FUnsent:1 Failed to load resource: the server responded with a status of 404 (Not Found)
recipes/undefined/tiddlers/%24%3A%2FWikiSettings%2Fsplit%2Fws-server:1 Failed to load resource: the server responded with a status of 404 (Not Found)
recipes/undefined/tiddlers/%24%3A%2FWikiSettings%2Fsplit%2Fwikis:1 Failed to load resource: the server responded with a status of 404 (Not Found)
recipes/undefined/tiddlers/%24%3A%2FWikiSettings:1 Failed to load resource: the server responded with a status of 404 (Not Found)
recipes/undefined/tiddlers/%24%3A%2FServerIP:1 Failed to load resource: the server responded with a status of 404 (Not Found)
recipes/undefined/tiddlers/%24%3A%2FEditionsList:1 Failed to load resource: the server responded with a status of 404 (Not Found)
recipes/undefined/tiddlers/%24%3A%2FBob%2FVisibleConfigurationTabs:1 Failed to load resource: the server responded with a status of 404 (Not Found)
recipes/undefined/tiddlers/%24%3A%2FBob%2FActiveLanguagesList:1 Failed to load resource: the server responded with a status of 404 (Not Found)
recipes/undefined/tiddlers/%24%3A%2FBob%2FActiveThemesList:1 Failed to load resource: the server responded with a status of 404 (Not Found)
recipes/undefined/tiddlers/%24%3A%2FBob%2FActivePluginList:1 Failed to load resource: the server responded with a status of 404 (Not Found)
recipes/undefined/tiddlers/%24%3A%2FWikiSettings%2Fsplit:1 Failed to load resource: the server responded with a status of 404 (Not Found)
recipes/undefined/tiddlers/%24%3A%2FWikiSettings%2Fsplit%2FexcludePluginList:1 Failed to load resource: the server responded with a status of 404 (Not Found)
recipes/undefined/tiddlers/%24%3A%2FWikiSettings%2Fsplit%2FincludePluginList:1 Failed to load resource: the server responded with a status of 404 (Not Found)
recipes/undefined/tiddlers/%24%3A%2FWikiSettings%2Fsplit%2FserverInfo:1 Failed to load resource: the server responded with a status of 404 (Not Found)
recipes/undefined/tiddlers/%24%3A%2FWikiSettings%2Fsplit%2FAPI:1 Failed to load resource: the server responded with a status of 404 (Not Found)
```

Other alerts were continuosly raised up while browsing the wiki. The behaviour did not change moving the wiki into `Bob/Wikis` folder. The following workaround solves:

1. I copy my plugins into `Bob/Plugins` folder, ... 
2. verify that plugins are correctly listed in "_Control Panel/Bob Settings/Manage Plugins_" tab, ...
3. stop old node.js instance, ...
4. edit _mywiki_'s `tiddlywiki.info` and remove all of the plugins from the plugins section, ...
5. update Bob settings with the path of my existing wiki, ...
6. restart Bob, ...
7. open _mywiki_ with the url provided by Bob (`127.0.0.1:8080/mywiki`), with no plugins, ...
8. navigate to "_Control Panel/Bob Settings/Manage Plugins_" in _mywiki_, ...
9. update plugin list, ...
10. select the needed plugins from the newly generated list, ...
11. save plugin selection and refresh page. Done.

Hope this can help. Regards,

)+(auloop

P.S.: recently I opened an issue in the BobExe repo, for a generic Bob related issue, despite all of your explanation posts. I noticed it just now. Thanks for having it answered as well ;^)))

---

I am running (check any that apply, put an x inside the [ ] to check a box, like this: [x]):

- [x] Windows
- [ ] OSX
- [ ] Linux
- [ ] Other

and using

- [ ] The nodejs version
- [x] The single file executable

Before posting I read [issue guidelines](./IssueGuidelines.md) and:

- [x] I am using the newest version
- [x] The answer to my question isn't listed in the documentation or this isn't
  a question
- [x] This is not a duplicate issue
- [x] I have not done anything that required me to set acceptance to
  `I Will Not Get Tech Support For This`

**Comment by inmysocks** [comment]
Thank you for the report, and I am glad that you found a work around.

One thing that may cause trouble is having two servers using the same wiki at the same time. Bob and the normal node server may interfere with each other so having both of them running with the same wiki can cause errors. I don't think that is the case here but it is possible.

The errors that you listed come from having the plugins that are used with the normal node server. Bob should be blocking them from being used so it is a bug that they are giving you trouble. The most likely reason is that there is a tiddler file for the plugins in your tiddlers folder.

I will check the code to see if I can find any errors that would allow those plugins to be active while using Bob.

**Comment by mauloop** [comment]
> One thing that may cause trouble is having two servers using the same wiki at the same time.

As from step 3, I shut the standard node.js server before starting Bob. I reported just the basic steps and omitted the boring details, but it took hours to run exhaustive tests. It also implied working on a copy of the original wiki. I cloned and destroyed it many timed and it was never served by two engines at the same time.

> The most likely reason is that there is a tiddler file for the plugins in your tiddlers folder.

No, I had no plugins tiddler in my tiddler folder. As I said I checked it many times and tested with other wikis too.

On the other hand, I did not have to change the tiddler folder content to make it work, but the `tiddlywiki.info`. The key was to remove the plugins from `tiddlywiki.info` and subsequently re-add them by the Bob's _Manage Plugin_ interface. I cannot explain why.

BTW, I also tried to not modify the `tiddlywiki.info`, but just to ignore the alerts and try to arrange with the _Manage Plugin_ interface to make it match the content of `tiddlywiki.info`. It did not success. The only working flow I found was the one I described in my initial post.

**Comment by inmysocks** [comment]
I finally ran into this problem as well and was able to fix what was causing it for me at least, the fix will be in version 1.2.2 so let me know if the problem continues for you after I get that released.

**Comment by mauloop** [comment]
Hi, Jed.

I played again the steps which made Bob to raise the XMLHttpRequest errors, as from my first post. This time everything looks to be working correctly. The plugins listed in `tiddlywiki.info` were automatically  recognized without the need of any workaround. I noticed some strange behaviours.

1.  

> With Bob 1.2.1, the _BobSettings_ gui in _ControlPanel_, showed up just two tabs in my imported wikis ("_Wiki Listing_" and "_Manage Plugins_"), while in the IndexWiki much more were available ("_Manual Settings_", "_Server Action_", ...). Now, with Bob 1.2.2, _ControlPanel/BobSetting_ displays all of the tabs as in the IndexWiki and below that the "_Wiki Listing_" and "_Manage Plugins_" tabs are displayed again.

2.  (this one happens with Bob 1.2.1 and 1.2.2 as well)

> I use a custom `$:/config/FileSystemPaths`, with this content:
> 
> ```
> [removeprefix[$:/_mau/]addprefix[_mau/]]
> [removeprefix[$:/]addprefix[system/]]
> [!prefix[$:/]addprefix[user/]]
> ```
> 
>  In my wiki I see that the following **non shadow** tiddlers have been crerated:
> 
> ```
>  $:/Bob/ActiveLanguagesList
>  $:/Bob/ActivePluginList
>  $:/Bob/ActiveThemesList
>  $:/Bob/AvailablePluginList
>  $:/Bob/VisibleConfigurationTabs
> ```
> In the tiddler folder I cannot find these, but just `$:/Bob/AvailablePluginList`. This one has been saved as:
> 
> ```
> tiddlers\system\Bob\system\Bob\AvailablePluginList.tid
> ```
> which is not what I expected, given the settings in `FileSystemPaths` (`system\bob` is repeated, nested in `system\bob` !!!)
> 
> These messages were displayed in the Bob's launch console, when I forced the plugin list update and `$:/Bob/AvailablePluginList` was first created.
> 
> ```
> NodeSettings - Parsed raw settings.
> Added route /^\/EditionM\/?$/
> TiddlyWiki version 5.1.19 with Bob version 1.2.2
> Serving on 127.0.0.1:8080
> (press ctrl-C to exit)
>  syncer-server-WebsocketAdaptor: Dispatching 'save' task: $:/StoryList
> new connection
> new connection
> For $:/Bob/AvailablePluginList, type is application/x-tiddler hasMetaFile is false filepath is Z:\opt64\Bob122\Wikis\EditionM\tiddlers\system\Bob\AvailablePluginList.tid
> saved file Z:\opt64\Bob122\Wikis\EditionM\tiddlers\system\Bob\AvailablePluginList.tid
> saved file Z:\opt64\Bob122\Wikis\EditionM\tiddlers\system\Bob\system\Bob\AvailablePluginList.tid
> saved file Z:\opt64\Bob122\Wikis\EditionM\tiddlers\system\Bob\system\Bob\AvailablePluginList.tid
> ```
>
>I have no idea of where all of others `$:/Bob/...` tiddlers are.

I'm still observing Bob 1.2.2 behaviour and I will report it to you if I noticed something else strange or misbehaving. Generally speaking I think Bob moved forward. A little step for a man, a great step for the Tiddlywiki community ;^))). Thanks for your help.

**Comment by inmysocks** [comment]
First, I am dyslexic, please don't add text decoration (bond, italic, quote blocks, different text colors, etc.) unless absolutely necessary, otherwise it is hard for me to read.

Second, you should list new issues in a new issue, please don't add unrelated problems to an existing issue.

And here are responses to what you listed:

1. That bug was fixed, make sure you are using the newest version. Currently it is 1.2.2r4, if you are using the newest version the most likely problem is that you have edited shadow tiddlers, check to see if you have a tiddler called $:/plugins/OokTech/Bob/AvailableWikisTab or $:/plugins/OokTech/Bob/Wiki Listing and delete them if they are present. After this you will have to check the 'Show available wikis in sidebar' checkbox again if you want wikis listed in the sidebar.

2. Bob makes tiddlers that aren't saved to the file system, the tiddlers you listed are not supposed bo be saved to the file system and are behaving correctly.
The paths part where it adds \system\Bob\system\Bob sounds like another bug I thought I had fixed. Please make a new issue for it.

**Comment by mauloop** [comment]
I will open a new issue as you suggested. I think this one is solved.
